src/Windows/Avalonia.Win32/Composition/CompositionConnector.cs

```diff
@@ -0,0 +1,143 @@
+        public ICompositionDrawingSurfaceInterop InitialiseWindowCompositionTree(IntPtr hwnd, out Windows.UI.Composition.Visual surfaceVisual, out IBlurHost blurHost)
+        {
+            var target = CreateDesktopWindowTarget(hwnd);
+
+            var surface = _graphicsDevice.CreateDrawingSurface(new Windows.Foundation.Size(0, 0),
+                Windows.Graphics.DirectX.DirectXPixelFormat.B8G8R8A8UIntNormalized,
+                Windows.Graphics.DirectX.DirectXAlphaMode.Premultiplied);
+
+            var surfaceInterop = surface.As<ICompositionDrawingSurfaceInterop>();
+
+            var brush = _compositor.CreateSurfaceBrush(surface);
+            var visual = _compositor.CreateSpriteVisual();
+
+            visual.Brush = brush;
+            visual.RelativeSizeAdjustment = new System.Numerics.Vector2(1, 1);
+
+            var container = _compositor.CreateContainerVisual();
+
+            target.Root = container;
+
+            var blur = CreateBlur();
+
+            blurHost = new CompositionBlurHost(blur);
+
+            container.Children.InsertAtTop(blur);
+
+            container.Children.InsertAtTop(visual);
+
+            visual.CompositeMode = CompositionCompositeMode.SourceOver;
+
+            surfaceVisual = container;
+
+            return surfaceInterop;
+        }
+
+        private SpriteVisual CreateBlur()
+        {
+            var blurEffect = new GaussianBlurEffect(new CompositionEffectSourceParameter("backdrop"));
+            var blurEffectFactory = _compositor.CreateEffectFactory(blurEffect);
+
+            var blurBrush = blurEffectFactory.CreateBrush();
+            var backDropBrush = _compositor.CreateBackdropBrush();
+
+            blurBrush.SetSourceParameter("backdrop", backDropBrush);
+
+            var saturateEffect = new SaturationEffect(blurEffect);
+            var satEffectFactory = _compositor.CreateEffectFactory(saturateEffect);
+
+            var satBrush = satEffectFactory.CreateBrush();
+            satBrush.SetSourceParameter("backdrop", backDropBrush);
+
+            var visual = _compositor.CreateSpriteVisual();
+            visual.IsVisible = false;
+
+            visual.RelativeSizeAdjustment = new System.Numerics.Vector2(1.0f, 1.0f);
+            visual.Brush = satBrush;
+
+            return visual;
+        }
+
+        private CompositionTarget CreateDesktopWindowTarget(IntPtr window)
+        {
+            var interop = _compositor.As<global::Windows.UI.Composition.Desktop.ICompositorDesktopInterop>();
+
+            interop.CreateDesktopWindowTarget(window, false, out var windowTarget);
+            return Windows.UI.Composition.Desktop.DesktopWindowTarget.FromAbi(windowTarget);
+        }
+
+        private void EnsureDispatcherQueue()
+        {
+            if (_dispatcherQueueController == null)
+            {
+                DispatcherQueueOptions options = new DispatcherQueueOptions();
+                options.apartmentType = DISPATCHERQUEUE_THREAD_APARTMENTTYPE.DQTAT_COM_STA;
```

Comment:
STA is NOT compatible with calls from multiple threads

Reply:
fixed.